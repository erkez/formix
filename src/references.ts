import { List, Seq } from 'immutable';

import FieldStateImpl from './FieldState';
import { ArrayFieldRef, FieldRef, FieldState, GenericFieldRef } from './types';

class FieldRefImpl<V> implements FieldRef<V> {
    readonly initialState: FieldState<V>;

    constructor(readonly initialValue: V, initialDisabled: boolean) {
        this.initialState = new FieldStateImpl(initialValue, initialDisabled, false, false);
    }
}

export function defineField<V>(initialValue: V, initialDisabled = false): FieldRef<V> {
    return new FieldRefImpl(initialValue, initialDisabled);
}

export function isFieldRef(field: unknown): field is FieldRef<unknown> {
    return field instanceof FieldRefImpl;
}

class ArrayFieldRefImpl<T extends GenericFieldRef, V>
    extends FieldRefImpl<List<T>>
    implements ArrayFieldRef<T, V> {
    constructor(
        initialItems: List<T>,
        readonly itemTemplate: (value: V) => T,
        initialDisabled: boolean
    ) {
        super(initialItems, initialDisabled);
    }
}

export function defineArrayField<T extends GenericFieldRef, V>(
    initialItems: Iterable<V>,
    itemTemplate: (value: V) => T,
    initialDisabled = false
): ArrayFieldRef<T, V> {
    return new ArrayFieldRefImpl(
        Seq(initialItems).map(itemTemplate).toList(),
        itemTemplate,
        initialDisabled
    );
}

export function isArrayFieldRef(field: unknown): field is ArrayFieldRef<GenericFieldRef, unknown> {
    return field instanceof ArrayFieldRefImpl;
}
