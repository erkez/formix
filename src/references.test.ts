import { defineArrayField, defineField, isArrayFieldRef, isFieldRef } from './references';

describe('references', () => {
    describe('FieldRef', () => {
        const fieldRef = defineField('fizz');

        it('should define FieldRef', () => {
            expect(fieldRef.initialState).toMatchSnapshot();
        });

        it('should define FieldRef initially disabled', () => {
            const disabledFieldRef = defineField('', true);
            expect(disabledFieldRef.initialState.disabled).toBe(true);
        });

        it('should correctly identify FieldRef', () => {
            expect(isFieldRef(fieldRef)).toBe(true);
            expect(isArrayFieldRef(fieldRef)).toBe(false);
        });
    });

    describe('ArrayFieldRef', () => {
        const arrayFieldRef = defineArrayField(['fizz'], (value) => ({ name: defineField(value) }));

        it('should define ArrayFieldRef', () => {
            expect(arrayFieldRef.initialState).toMatchSnapshot();
        });

        it('should define ArrayFieldRef initially disabled', () => {
            const disabledFieldRef = defineArrayField([], () => ({}), true);
            expect(disabledFieldRef.initialState.disabled).toBe(true);
        });

        it('should create item in array', () => {
            const item = arrayFieldRef.itemTemplate('buzz');
            expect(item).toEqual({
                name: defineField('buzz')
            });
        });

        it('should correctly identify ArrayFieldRef', () => {
            expect(isArrayFieldRef(arrayFieldRef)).toBe(true);
            expect(isFieldRef(arrayFieldRef)).toBe(true);
        });
    });
});
