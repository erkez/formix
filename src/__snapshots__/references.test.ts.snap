// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`references ArrayFieldRef should define ArrayFieldRef 1`] = `
FieldState {
  "active": false,
  "disabled": false,
  "touched": false,
  "value": Immutable.List [
    Object {
      "name": FieldRefImpl {
        "initialState": FieldState {
          "active": false,
          "disabled": false,
          "touched": false,
          "value": "fizz",
        },
      },
    },
  ],
}
`;

exports[`references FieldRef should define FieldRef 1`] = `
FieldState {
  "active": false,
  "disabled": false,
  "touched": false,
  "value": "fizz",
}
`;
